core/drand_beacon_control.go

```diff
@@ -1145,15 +1144,111 @@ func (bp *BeaconProcess) StartFollowChain(req *drand.StartFollowRequest, stream
 		Clock:    bp.opts.clock,
 		NodeAddr: bp.priv.Public.Address(),
 	})
-	go syncer.Run()
+	go syncer.Run(ctx)
 	defer syncer.Stop()
-	syncer.RequestSync(peers, req.GetUpTo())
+
+	for {
+		syncer.RequestSync(0, req.GetUpTo(), peers)
+		// wait for all the callbacks to be called and progress sent before returning
+		select {
+		case <-done:
+			return nil
+		case <-ctx.Done():
+			return ctx.Err()
+		// we re-send a sync request every 10 periods in case the process got staled and let the sync manager handle
+		case <-time.After(info.Period * 10): //nolint:gomnd
```

Comment:
is this going to end up leaking channels that never get drained?

Reply:
`time.After` is GC after it fires, so I'd say it's acceptable, no? If we want to have better control on it, I could setup my own timer as well